text/2018-11-24-consistent-replica-read.md

```diff
@@ -0,0 +1,89 @@
+# Consistent Replica Read
+
+## Summary
+
+*Consistent Replica Read* supports consistent read on the follower replica or
+learner replica, which can reduce the read load on the leader. *Consistent
+Replica Read* is also useful in OLAP or backup scenarios.
+
+The key to achieve consistent read on the replica is `ReadIndex`, which was
+introduced in Chapter 6 of the [Raft
+thesis](https://ramcloud.stanford.edu/~ongaro/thesis.pdf).
+
+## Motivation
+
+Currently, all read requests are handled on the leader, and load balancing is
+performed only by the `balanceLeader` operator, which have lead to many
+limitations. On one hand, we want a way to improve read throughput and divert
+load away from the leader; on the other hand, we want some nodes to handle
+read-only service and some of the heavy analytics queries.
+
+## Detail design
+
+### Replica Read
+
+As indicated in the Summary section, the key to implementing *Consistent
+Replica Read* is `ReadIndex`, which has already been implemented in our
+[raft-rs](http://github.com/pingcap/raft-rs) library. For now, some read
+requests are already handled using the `ReadIndex` policy. It follows these
+steps:
+
+1. Determine the policy of the read request. If the request is required to
+   check the quorum, the replica is a leader with an expired lease, or the
+   replica is a follower or a learner, it should be using the `ReadIndex`
+   policy.
+2. Get the latest `CommitIndex` from the raft library. Assign a `uuid` for the
+   request and invoke the `read_index` API of the raft library with the data
+   that was encoded by `uuid`. Then the raft state machine will notify our apply
+   state machine with the `read_status` through the `Ready` message. And we can
```

Comment:
```suggestion
   state machine with `read_status` through the `Ready` message. And we can
```

Reply:
We would say the `read_status` parameter (variable, option)
or just `read_status` to indicate the same.